File: backend/app/schemas/estimate_workbench.py
"""Pydantic schemas for the estimate workbench API."""

from __future__ import annotations

from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field

from app.schemas.base import UTCModel


# ---------------------------------------------------------------------------
# Recalc (stateless)
# ---------------------------------------------------------------------------


class FabLineRecalcIn(BaseModel):
    material: str = ""
    qty: int = 1
    thickness_in: Optional[float] = None
    width_in: Optional[float] = None
    length_in: Optional[float] = None
    cut_length_in: Optional[float] = None
    pierce_count: int = 0
    bend_count: int = 0
    weld_length_in: Optional[float] = None
    weld_minutes_ea: Optional[float] = None
    material_family_override: Optional[str] = None  # mild|stainless|aluminum
    include_material: bool = True
    include_laser: bool = True
    include_brake: bool = True
    include_weld: bool = True
    price_per_lb: float = 0.0
    density_lb_per_in3: float = 0.284
    detail_name: Optional[str] = None
    part_number: Optional[str] = None


class BuyoutLineRecalcIn(BaseModel):
    qty: float = 1.0
    unit_cost: float = 0.0
    description: str = ""
    category: Optional[str] = None
    vendor: Optional[str] = None
    part_number: Optional[str] = None
    part_id: Optional[int] = None
    price_source: Optional[str] = None
    confidence: Optional[str] = None
    verification_note: Optional[str] = None


class MachinedLineRecalcIn(BaseModel):
    material: str = ""
    qty: int = 1
    stock_dia_in: Optional[float] = None
    stock_length_in: Optional[float] = None
    turning_minutes: float = 0.0
    milling_minutes: float = 0.0
    price_per_lb: float = 0.0
    density_lb_per_in3: float = 0.284
    description: Optional[str] = None
    part_number: Optional[str] = None
    confidence: Optional[str] = None
    verification_note: Optional[str] = None


class AssemblyRecalcIn(BaseModel):
    name: str = "Assembly"
    assembly_labor_hrs: float = 0.0
    electrical_labor_hrs: float = 0.0
    notes: Optional[str] = None
    sort_order: Optional[int] = None
    fab_lines: List[FabLineRecalcIn] = Field(default_factory=list)
    buyout_lines: List[BuyoutLineRecalcIn] = Field(default_factory=list)


class RatesOverrideIn(BaseModel):
    """Optional overrides; missing fields fall back to DB / Excel defaults."""

    laser_rate: Optional[float] = None
    brake_rate: Optional[float] = None
    weld_rate: Optional[float] = None
    scrap_factor: Optional[float] = None
    laser_speed_fallback: Optional[float] = None
    pierce_time_fallback: Optional[float] = None
    target_margin: Optional[float] = None
    overhead_markup: Optional[float] = None
    buyout_markup: Optional[float] = None
    consumables_per_job: Optional[float] = None
    assembly_labor_rate: Optional[float] = None
    electrical_labor_rate: Optional[float] = None
    cnc_turning_rate: Optional[float] = None
    cnc_mill_rate: Optional[float] = None


class RecalcRequest(BaseModel):
    assemblies: List[AssemblyRecalcIn] = Field(default_factory=list)
    machined_parts: List[MachinedLineRecalcIn] = Field(default_factory=list)
    rates: Optional[RatesOverrideIn] = None


class CalcMessageOut(BaseModel):
    code: str
    message: str
    field: Optional[str] = None
    suggested_value: Optional[float] = None


class FabLineRecalcOut(BaseModel):
    detail_name: Optional[str] = None
    part_number: Optional[str] = None
    material_family: str
    weight_ea_lb: float
    material_cost: float
    laser_cost: float
    laser_hours: float
    brake_cost: float
    brake_hours: float
    weld_cost: float
    weld_hours: float
    weld_minutes_ea: float
    line_total: float
    cut_length_used: float
    errors: List[CalcMessageOut] = Field(default_factory=list)
    warnings: List[CalcMessageOut] = Field(default_factory=list)


class MachinedLineRecalcOut(BaseModel):
    description: Optional[str] = None
    weight_ea_lb: float
    material_cost: float
    turning_cost: float
    turning_hours: float
    milling_cost: float
    milling_hours: float
    line_total: float


class BidSummaryOut(BaseModel):
    fab_material: float
    fab_laser: float
    fab_brake: float
    fab_weld: float
    fab_subtotal: float
    buyout_subtotal: float
    buyout_marked_up: float
    assembly_labor_cost: float
    electrical_labor_cost: float
    machined_subtotal: float
    laser_hours: float
    brake_hours: float
    weld_hours: float
    assembly_hours: float
    electrical_hours: float
    subtotal_before_oh: float
    overhead: float
    consumables: float
    cogs: float
    sell_price: float
    target_margin: float
    errors: List[CalcMessageOut] = Field(default_factory=list)


class RecalcResponse(BaseModel):
    fab_lines: List[FabLineRecalcOut]
    machined_parts: List[MachinedLineRecalcOut]
    bid_summary: BidSummaryOut
    shop_data_source: str = "defaults"


# ---------------------------------------------------------------------------
# Persist tree (GET / PUT / POST create)
# ---------------------------------------------------------------------------


class FabLinePersistIn(FabLineRecalcIn):
    sort_order: Optional[int] = None
    confidence: Optional[str] = None
    verification_note: Optional[str] = None


class BuyoutLinePersistIn(BuyoutLineRecalcIn):
    sort_order: Optional[int] = None


class MachinedLinePersistIn(MachinedLineRecalcIn):
    sort_order: Optional[int] = None


class AssemblyPersistIn(BaseModel):
    name: str = "Assembly"
    assembly_labor_hrs: float = 0.0
    electrical_labor_hrs: float = 0.0
    notes: Optional[str] = None
    sort_order: Optional[int] = None
    fab_lines: List[FabLinePersistIn] = Field(default_factory=list)
    buyout_lines: List[BuyoutLinePersistIn] = Field(default_factory=list)


class WorkbenchSaveRequest(BaseModel):
    version: int = Field(..., description="Optimistic lock — must match current estimate.version")
    assemblies: List[AssemblyPersistIn] = Field(default_factory=list)
    machined_parts: List[MachinedLinePersistIn] = Field(default_factory=list)


class WorkbenchCreateRequest(BaseModel):
    rfq_package_id: int


class FabLineOut(UTCModel):
    id: int
    sort_order: int
    part_number: Optional[str] = None
    detail_name: str
    material: str
    material_family_override: Optional[str] = None
    qty: int
    thickness_in: Optional[float] = None
    width_in: Optional[float] = None
    length_in: Optional[float] = None
    cut_length_in: Optional[float] = None
    pierce_count: int
    bend_count: int
    weld_length_in: Optional[float] = None
    weld_minutes_ea: Optional[float] = None
    include_material: bool
    include_laser: bool
    include_brake: bool
    include_weld: bool
    weight_ea_lb: Optional[float] = None
    material_cost: float
    laser_cost: float
    laser_hours: float
    brake_cost: float
    brake_hours: float
    weld_cost: float
    weld_hours: float
    line_total: float
    calc_warnings: Optional[List[Dict[str, Any]]] = None
    calc_errors: Optional[List[Dict[str, Any]]] = None
    confidence: str
    verification_note: Optional[str] = None
    version: int


class BuyoutLineOut(UTCModel):
    id: int
    sort_order: int
    category: Optional[str] = None
    vendor: Optional[str] = None
    part_number: Optional[str] = None
    part_id: Optional[int] = None
    description: str
    qty: float
    unit_cost: float
    extended_cost: float
    price_source: Optional[str] = None
    confidence: str
    verification_note: Optional[str] = None
    version: int


class MachinedLineOut(UTCModel):
    id: int
    sort_order: int
    part_number: Optional[str] = None
    description: str
    material: str
    qty: int
    stock_dia_in: Optional[float] = None
    stock_length_in: Optional[float] = None
    turning_minutes: float
    milling_minutes: float
    weight_ea_lb: Optional[float] = None
    material_cost: float
    turning_cost: float
    turning_hours: float
    milling_cost: float
    milling_hours: float
    line_total: float
    confidence: str
    verification_note: Optional[str] = None
    version: int


class AssemblyOut(UTCModel):
    id: int
    name: str
    sort_order: int
    assembly_labor_hrs: float
    electrical_labor_hrs: float
    notes: Optional[str] = None
    version: int
    fab_lines: List[FabLineOut] = Field(default_factory=list)
    buyout_lines: List[BuyoutLineOut] = Field(default_factory=list)


class WorkbenchResponse(UTCModel):
    estimate_id: int
    rfq_package_id: int
    quote_id: Optional[int] = None
    version: int
    currency: str
    grand_total: float
    material_total: float
    hardware_consumables_total: float
    shop_labor_oh_total: float
    margin_total: float
    internal_breakdown: Optional[Dict[str, Any]] = None
    assemblies: List[AssemblyOut] = Field(default_factory=list)
    machined_parts: List[MachinedLineOut] = Field(default_factory=list)
    shop_data_source: Optional[str] = None
    verification: Optional["VerificationReportOut"] = None


# ---------------------------------------------------------------------------
# Phase 3 — Verification + finalize
# ---------------------------------------------------------------------------


class PriorityActionOut(BaseModel):
    category: str
    line_id: int
    assembly_id: Optional[int] = None
    assembly_name: Optional[str] = None
    label: str
    confidence: str
    reason: str
    anchor: str
    line_total: float = 0.0


class CategorySummaryOut(BaseModel):
    label: str
    total: float
    count: int
    confirmed: int
    majority: int
    review: int


class VerificationReportOut(BaseModel):
    estimate_id: int
    status: str
    can_finalize: bool
    review_count: int
    blocker_count: int
    categories: List[CategorySummaryOut]
    priority_actions: List[PriorityActionOut]
    blockers: List[Dict[str, Any]] = Field(default_factory=list)
    banner: Optional[str] = None


class FinalizeRequest(BaseModel):
    valid_days: int = 30
    force: bool = False  # admin escape hatch — still audited


class FinalizeResponse(BaseModel):
    estimate_id: int
    quote_id: int
    quote_number: str
    grand_total: float
    forced: bool = False
    verification: VerificationReportOut


# ---------------------------------------------------------------------------
# Phase 4 — PDF / RFQ extraction assist
# ---------------------------------------------------------------------------


class ExtractFromRfqRequest(BaseModel):
    rfq_package_id: Optional[int] = None  # default: estimate's package
    use_llm: bool = True
    apply: bool = False  # if True, replace workbench tree with draft
    replace: bool = True  # when apply: replace vs merge into first assembly
    version: Optional[int] = None  # required when apply=True


class ExtractionSummaryOut(BaseModel):
    fab_count: int = 0
    buyout_count: int = 0
    review_count: int = 0
    majority_count: int = 0
    confirmed_count: int = 0


class FabLineDraftOut(BaseModel):
    """Pre-persist fab line from extraction (no DB id yet)."""

    detail_name: str = "Detail"
    part_number: Optional[str] = None
    material: str = ""
    qty: int = 1
    thickness_in: Optional[float] = None
    width_in: Optional[float] = None
    length_in: Optional[float] = None
    cut_length_in: Optional[float] = None
    pierce_count: int = 0
    bend_count: int = 0
    weld_length_in: Optional[float] = None
    include_material: bool = True
    include_laser: bool = True
    include_brake: bool = True
    include_weld: bool = False
    confidence: Optional[str] = None
    verification_note: Optional[str] = None
    field_confidence: Optional[Dict[str, Any]] = None


class BuyoutLineDraftOut(BaseModel):
    description: str = "Buyout"
    part_number: Optional[str] = None
    category: Optional[str] = None
    vendor: Optional[str] = None
    qty: float = 1.0
    unit_cost: float = 0.0
    price_source: Optional[str] = None
    confidence: Optional[str] = None
    verification_note: Optional[str] = None
    field_confidence: Optional[Dict[str, Any]] = None


class AssemblyDraftOut(BaseModel):
    name: str = "Assembly"
    sort_order: int = 0
    assembly_labor_hrs: float = 0.0
    electrical_labor_hrs: float = 0.0
    fab_lines: List[FabLineDraftOut] = Field(default_factory=list)
    buyout_lines: List[BuyoutLineDraftOut] = Field(default_factory=list)


class ExtractFromRfqResponse(BaseModel):
    mode: str
    assemblies: List[AssemblyDraftOut] = Field(default_factory=list)
    machined_parts: List[Dict[str, Any]] = Field(default_factory=list)
    summary: ExtractionSummaryOut
    warnings: List[str] = Field(default_factory=list)
    applied: bool = False
    workbench: Optional[WorkbenchResponse] = None
    extraction_artifact: Optional[Dict[str, Any]] = None


# ---------------------------------------------------------------------------
# Phase 5 — Shop Data + quoted-vs-actual
# ---------------------------------------------------------------------------


class CutBendRowOut(BaseModel):
    id: int
    table_id: int
    sort_order: int
    thickness_in: Optional[float] = None
    gauge: Optional[int] = None
    mild_steel: Optional[float] = None
    stainless: Optional[float] = None
    aluminum: Optional[float] = None
    value: Optional[float] = None
    fillet_leg_in: Optional[float] = None
    arc_in_per_min: Optional[float] = None
    min_per_in: Optional[float] = None
    notes: Optional[str] = None


class CutBendTableOut(BaseModel):
    id: int
    kind: str
    name: str
    description: Optional[str] = None
    columns: List[str] = Field(default_factory=list)
    rows: List[CutBendRowOut] = Field(default_factory=list)
    updated_at: Optional[str] = None


class ShopDataTablesResponse(BaseModel):
    tables: List[CutBendTableOut]
    source: str = "db"


class CutBendRowUpdateRequest(BaseModel):
    note: str = Field(..., min_length=1, description="Required: why this cell changed")
    thickness_in: Optional[float] = None
    gauge: Optional[int] = None
    mild_steel: Optional[float] = None
    stainless: Optional[float] = None
    aluminum: Optional[float] = None
    value: Optional[float] = None
    fillet_leg_in: Optional[float] = None
    arc_in_per_min: Optional[float] = None
    min_per_in: Optional[float] = None
    notes: Optional[str] = None


class CutBendRowCreateRequest(BaseModel):
    note: str = Field(..., min_length=1)
    thickness_in: Optional[float] = None
    gauge: Optional[int] = None
    mild_steel: Optional[float] = None
    stainless: Optional[float] = None
    aluminum: Optional[float] = None
    value: Optional[float] = None
    fillet_leg_in: Optional[float] = None
    arc_in_per_min: Optional[float] = None
    min_per_in: Optional[float] = None
    notes: Optional[str] = None


class ShopDataHistoryItemOut(BaseModel):
    id: int
    entity_id: Optional[int] = None
    entity_name: Optional[str] = None
    action: str
    field_changed: Optional[str] = None
    old_value: Optional[Any] = None
    new_value: Optional[Any] = None
    note: Optional[str] = None
    changed_by: Optional[int] = None
    changed_at: Optional[str] = None


class JobActualOut(BaseModel):
    id: int
    quote_estimate_id: Optional[int] = None
    work_order_id: Optional[int] = None
    job_label: Optional[str] = None
    quoted_laser_hours: float = 0.0
    quoted_brake_hours: float = 0.0
    quoted_weld_hours: float = 0.0
    actual_laser_hours: Optional[float] = None
    actual_brake_hours: Optional[float] = None
    actual_weld_hours: Optional[float] = None
    delta_laser_pct: Optional[float] = None
    delta_brake_pct: Optional[float] = None
    delta_weld_pct: Optional[float] = None
    notes: Optional[str] = None
    entered_by: Optional[int] = None
    created_at: Optional[str] = None
    updated_at: Optional[str] = None
    propose_tune: List[Dict[str, Any]] = Field(default_factory=list)


class JobActualUpsertRequest(BaseModel):
    quote_estimate_id: Optional[int] = None
    work_order_id: Optional[int] = None
    job_label: Optional[str] = None
    actual_laser_hours: Optional[float] = None
    actual_brake_hours: Optional[float] = None
    actual_weld_hours: Optional[float] = None
    quoted_laser_hours: Optional[float] = None
    quoted_brake_hours: Optional[float] = None
    quoted_weld_hours: Optional[float] = None
    notes: Optional[str] = None

File: frontend/src/types/estimateWorkbench.ts
/** Types for the Estimate Workbench (Excel-replacement quoting). */

export type ConfidenceLevel = 'confirmed' | 'majority' | 'review';
export type MaterialFamily = 'mild' | 'stainless' | 'aluminum';

export interface CalcMessage {
  code: string;
  message: string;
  field?: string | null;
  suggested_value?: number | null;
}

export interface FabLineDraft {
  detail_name: string;
  part_number?: string | null;
  material: string;
  material_family_override?: MaterialFamily | string | null;
  qty: number;
  thickness_in?: number | null;
  width_in?: number | null;
  length_in?: number | null;
  cut_length_in?: number | null;
  pierce_count: number;
  bend_count: number;
  weld_length_in?: number | null;
  weld_minutes_ea?: number | null;
  include_material: boolean;
  include_laser: boolean;
  include_brake: boolean;
  include_weld: boolean;
  price_per_lb?: number;
  density_lb_per_in3?: number;
  confidence?: ConfidenceLevel | string;
  verification_note?: string | null;
  sort_order?: number;
  // Cached outputs (from server)
  id?: number;
  weight_ea_lb?: number | null;
  material_cost?: number;
  laser_cost?: number;
  laser_hours?: number;
  brake_cost?: number;
  brake_hours?: number;
  weld_cost?: number;
  weld_hours?: number;
  line_total?: number;
  calc_warnings?: CalcMessage[] | null;
  calc_errors?: CalcMessage[] | null;
  version?: number;
}

export interface BuyoutLineDraft {
  description: string;
  qty: number;
  unit_cost: number;
  category?: string | null;
  vendor?: string | null;
  part_number?: string | null;
  part_id?: number | null;
  price_source?: string | null;
  confidence?: ConfidenceLevel | string;
  verification_note?: string | null;
  sort_order?: number;
  id?: number;
  extended_cost?: number;
  version?: number;
}

export interface MachinedLineDraft {
  description: string;
  material: string;
  qty: number;
  part_number?: string | null;
  stock_dia_in?: number | null;
  stock_length_in?: number | null;
  turning_minutes: number;
  milling_minutes: number;
  price_per_lb?: number;
  density_lb_per_in3?: number;
  confidence?: ConfidenceLevel | string;
  verification_note?: string | null;
  sort_order?: number;
  id?: number;
  weight_ea_lb?: number | null;
  material_cost?: number;
  turning_cost?: number;
  turning_hours?: number;
  milling_cost?: number;
  milling_hours?: number;
  line_total?: number;
  version?: number;
}

export interface AssemblyDraft {
  name: string;
  assembly_labor_hrs: number;
  electrical_labor_hrs: number;
  notes?: string | null;
  sort_order?: number;
  fab_lines: FabLineDraft[];
  buyout_lines: BuyoutLineDraft[];
  id?: number;
  version?: number;
}

export interface BidSummary {
  fab_material: number;
  fab_laser: number;
  fab_brake: number;
  fab_weld: number;
  fab_subtotal: number;
  buyout_subtotal: number;
  buyout_marked_up: number;
  assembly_labor_cost: number;
  electrical_labor_cost: number;
  machined_subtotal: number;
  laser_hours: number;
  brake_hours: number;
  weld_hours: number;
  assembly_hours: number;
  electrical_hours: number;
  subtotal_before_oh: number;
  overhead: number;
  consumables: number;
  cogs: number;
  sell_price: number;
  target_margin: number;
  errors?: CalcMessage[];
}

export interface RecalcResponse {
  fab_lines: Array<{
    detail_name?: string | null;
    part_number?: string | null;
    material_family: string;
    weight_ea_lb: number;
    material_cost: number;
    laser_cost: number;
    laser_hours: number;
    brake_cost: number;
    brake_hours: number;
    weld_cost: number;
    weld_hours: number;
    weld_minutes_ea: number;
    line_total: number;
    cut_length_used: number;
    errors: CalcMessage[];
    warnings: CalcMessage[];
  }>;
  machined_parts: Array<{
    description?: string | null;
    weight_ea_lb: number;
    material_cost: number;
    turning_cost: number;
    turning_hours: number;
    milling_cost: number;
    milling_hours: number;
    line_total: number;
  }>;
  bid_summary: BidSummary;
  shop_data_source: string;
}

export interface WorkbenchResponse {
  estimate_id: number;
  rfq_package_id: number;
  quote_id?: number | null;
  version: number;
  currency: string;
  grand_total: number;
  material_total: number;
  hardware_consumables_total: number;
  shop_labor_oh_total: number;
  margin_total: number;
  internal_breakdown?: Record<string, number | string> | null;
  assemblies: Array<{
    id: number;
    name: string;
    sort_order: number;
    assembly_labor_hrs: number;
    electrical_labor_hrs: number;
    notes?: string | null;
    version: number;
    fab_lines: FabLineDraft[];
    buyout_lines: BuyoutLineDraft[];
  }>;
  machined_parts: MachinedLineDraft[];
  shop_data_source?: string | null;
  verification?: VerificationReport | null;
}

export interface PriorityAction {
  category: string;
  line_id: number;
  assembly_id?: number | null;
  assembly_name?: string | null;
  label: string;
  confidence: string;
  reason: string;
  anchor: string;
  line_total: number;
}

export interface CategorySummary {
  label: string;
  total: number;
  count: number;
  confirmed: number;
  majority: number;
  review: number;
}

export interface VerificationReport {
  estimate_id: number;
  status: string;
  can_finalize: boolean;
  review_count: number;
  blocker_count: number;
  categories: CategorySummary[];
  priority_actions: PriorityAction[];
  blockers: Array<Record<string, unknown>>;
  banner?: string | null;
}

export interface ExtractionSummary {
  fab_count: number;
  buyout_count: number;
  review_count: number;
  majority_count: number;
  confirmed_count: number;
}

export interface ExtractFromRfqResponse {
  mode: string;
  assemblies: AssemblyDraft[];
  machined_parts: MachinedLineDraft[];
  summary: ExtractionSummary;
  warnings: string[];
  applied: boolean;
  workbench?: WorkbenchResponse | null;
  extraction_artifact?: Record<string, unknown> | null;
}

export function emptyFabLine(index = 0): FabLineDraft {
  return {
    detail_name: `Detail ${index + 1}`,
    material: 'A36 Mild Steel',
    qty: 1,
    pierce_count: 0,
    bend_count: 0,
    include_material: true,
    include_laser: true,
    include_brake: true,
    include_weld: true,
    confidence: 'review',
    material_cost: 0,
    laser_cost: 0,
    laser_hours: 0,
    brake_cost: 0,
    brake_hours: 0,
    weld_cost: 0,
    weld_hours: 0,
    line_total: 0,
  };
}

export function emptyBuyoutLine(): BuyoutLineDraft {
  return {
    description: '',
    qty: 1,
    unit_cost: 0,
    confidence: 'review',
    extended_cost: 0,
  };
}

export function emptyMachinedLine(index = 0): MachinedLineDraft {
  return {
    description: `Machined ${index + 1}`,
    material: '1018 CD Bar',
    qty: 1,
    turning_minutes: 0,
    milling_minutes: 0,
    confidence: 'review',
    material_cost: 0,
    turning_cost: 0,
    turning_hours: 0,
    milling_cost: 0,
    milling_hours: 0,
    line_total: 0,
  };
}

export function emptyAssembly(index = 0): AssemblyDraft {
  return {
    name: `Assembly ${index + 1}`,
    assembly_labor_hrs: 0,
    electrical_labor_hrs: 0,
    fab_lines: [emptyFabLine(0)],
    buyout_lines: [],
  };
}

export function workbenchToDrafts(wb: WorkbenchResponse): {
  assemblies: AssemblyDraft[];
  machined_parts: MachinedLineDraft[];
} {
  return {
    assemblies: (wb.assemblies || []).map((a) => ({
      id: a.id,
      name: a.name,
      sort_order: a.sort_order,
      assembly_labor_hrs: a.assembly_labor_hrs,
      electrical_labor_hrs: a.electrical_labor_hrs,
      notes: a.notes,
      version: a.version,
      fab_lines: (a.fab_lines || []).map((f) => ({ ...f })),
      buyout_lines: (a.buyout_lines || []).map((b) => ({ ...b })),
    })),
    machined_parts: (wb.machined_parts || []).map((m) => ({ ...m })),
  };
}
